### PR TITLE
Resolution tracing

### PR DESCRIPTION
**Pull request by jmsfltchr**
## What is the goal of this PR?

We introduce a visual reasoner tracing tool that displays the internal actions of the reasoning engine. It maps out the messages passed between the actors of the message passing system that implements reasoning in Grakn. This has been built for internal use, but is exposed for use by anyone using an option on transactions.

## What are the changes implemented in this PR?

- Record messages sent between actors in the reasoner.
- Save these messages to `.dot` files, one for each request that is sent into the system.
- These `.dot` files can be used to generate `.svg` visuals using `graphviz`
- Due to the async nature of the system, requests need to be serialised to give clear traces, so `parallel` needs to be set to `false` on the transaction.
- This tracer needs to be enabled by setting `traceInference` on the transaction. This is separate from the database's log level.
- Requires https://github.com/graknlabs/protocol/pull/122

### CODE REVIEW

File: common/parameters/Options.java [hunk 1]
```diff
@@ -24,14 +24,16 @@
...
 import static grakn.core.common.exception.ErrorMessage.Reasoner.REASONING_CANNOT_BE_TOGGLED_PER_QUERY;
 
 public abstract class Options<PARENT extends Options<?, ?>, SELF extends Options<?, ?>> {
 
     public static final int DEFAULT_RESPONSE_BATCH_SIZE = 50;
     public static final int DEFAULT_SESSION_IDLE_TIMEOUT_MILLIS = 10_000;
     public static final int DEFAULT_SCHEMA_LOCK_ACQUIRE_TIMEOUT_MILLIS = 10_000;
-    public static final boolean DEFAULT_INFER = false;
+    public static final boolean DEFAULT_INFERENCE = false;
+    public static final boolean DEFAULT_INFERENCE_LOGGING = false;
```
Comment (flyingsilverfin) [inline common/parameters/Options.java:36]: should be called `inference_tracing` to match option

File: reasoner/ReasonerProducer.java [hunk 2]
```diff
@@ -72,10 +78,15 @@ public ReasonerProducer(Conjunction conjunction, ResolverRegistry resolverRegist
         this.processing = new AtomicInteger();
     }
 
-    public ReasonerProducer(Disjunction disjunction, ResolverRegistry resolverRegistry, GraqlMatch.Modifiers modifiers) {
+    public ReasonerProducer(Disjunction disjunction, ResolverRegistry resolverRegistry, GraqlMatch.Modifiers modifiers,
+                            Options.Query options) {
+        if (options.traceInference()) ResolutionTracer.initialiseOrGet(options.logsDir());
```
Comment (flyingsilverfin) [inline reasoner/ReasonerProducer.java:83]: `initialise()` and `get()` separate

File: reasoner/resolution/framework/ResolutionTracer.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+import java.nio.file.Path;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static grakn.core.common.exception.ErrorMessage.Internal.ILLEGAL_STATE;
+import static grakn.core.common.exception.ErrorMessage.Reasoner.REASONER_TRACING_CALL_TO_FINISH_BEFORE_START;
+import static grakn.core.common.exception.ErrorMessage.Reasoner.REASONER_TRACING_HAS_NOT_BEEN_INITIALISED;
+
+public final class ResolutionTracer {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ResolutionTracer.class);
```
Comment (flyingsilverfin) [inline reasoner/resolution/framework/ResolutionTracer.java:37]: do we need logging in the logger :D

File: test/integration/logic/TypeResolverTest.java [hunk 4]
```diff
@@ -57,16 +58,18 @@
...
     private static RocksSession session;
     private static RocksTransaction transaction;
 
     @BeforeClass
     public static void open_session() throws IOException {
-        Util.resetDirectory(directory);
-        grakn = RocksGrakn.open(directory);
+        Util.resetDirectory(dataDir);
+        grakn = RocksGrakn.open(options);
+        String database = "type-resolver-test";
```
Comment (flyingsilverfin) [inline test/integration/logic/TypeResolverTest.java:72]: in the other tests this is not in the beforeclass right?

File: test/integration/reasoner/ReiterationTest.java [hunk 5]
```diff
@@ -120,6 +124,7 @@ public void test_first_iteration_exhausts_and_second_iteration_recurses_infinite
                 int[] doneInIteration = {0};
                 boolean[] receivedInferredAnswer = {false};
 
+                ResolutionTracer.initialiseOrGet(logDir).start();
```
Comment (flyingsilverfin) [inline test/integration/reasoner/ReiterationTest.java:127]: the `initialise()` should be in the `setup()`

File: reasoner/ReasonerProducer.java [hunk 6]
```diff
@@ -47,8 +49,7 @@
 
     private static final Logger LOG = LoggerFactory.getLogger(ReasonerProducer.class);
 
-    private static final int COMPUTE_SIZE = Executors.PARALLELISATION_FACTOR * 2;
-
+    private int computeSize = Executors.PARALLELISATION_FACTOR * 2;
```
Comment (flyingsilverfin) [inline reasoner/ReasonerProducer.java:52]: we normally only initialise static members like this - everything else should be in the constructor

**Review by flyingsilverfin — APPROVED** [review]

**Review by haikalpribadi — APPROVED** [review]

**Pull request merged**